### PR TITLE
Adding JobListener to report Quartz metrics

### PR DESCRIPTION
**Pull request by comunacho**
On top of what we've added to publish instrumentation metrics using micrometer and statsd, this change adds two classes:

- A Quartz `JobListener` that runs upon completion of each job defined within the codebase and publishes timing metrics for each job's execution time (in milliseconds) in the `quartz.jobs.execution` timer, including `jobGroup` and `jobClass` tags. To enable this metric, the `l10n.management.metrics.quartz.enabled` property must be `true`.
- A `Scheduled` task, which collects "queue" sizes and publishes them as gauges in the `quartz.pending.jobs` gauge, including `jobGroup` and `jobClass` tags. This task runs only when MySQL is being used as the datasource. To enable this metric, the `l10n.management.metrics.quartz.queue-monitoring.enabled` property must be `true`. Also, a `l10n.management.metrics.quartz.queue-monitoring.execution-rate` property is provided to execute the task with a fixed period in milliseconds between invocations. It can be defined in milliseconds, of following a [Duration](https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html#parse-java.lang.CharSequence-) compliant string.

### CODE REVIEW

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzMetricsReportingJobListener.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    @Override
+    public String getName() {
+        return getClass().getSimpleName();
+    }
+
+    @Override
+    public void jobWasExecuted(JobExecutionContext context, JobExecutionException jobException) {
+        try {
+            registerExecutionTime(context);
+            registerPendingJobs(context);
```
[the remark this answers is not in the document]
Comment (comunacho) [reply webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzMetricsReportingJobListener.java:53]: I agree. I do have mixed feelings as well. As we discussed, the other option here would be to run this method during `jobToBeExecuted`, but at the end of the day, that'll have the exact same effect.

As we discussed, other options would be:

1) Using a `ScheduledTask` to run this query periodically (but we'll need to dedup and probably sanitize metrics being pushed through micrometer)
2) Using a Quartz job to pull these metrics within a scheduled job running in Quartz as well (which can lead to stale metrics or not having metrics at all if quartz is busy processing other jobs)
2.a) On top of this idea, we could play around with job priorities. But with slow & long-running jobs even priorities would create a bottleneck.
2.b) Instead of using priorities, we could create a dedicated scheduler on where we could execute jobs collecting metrics. This would ensure we would collect metrics periodically, but we would still using Quartz to collect metrics.

Another approach would encompass using

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzMetricsReportingJobListener.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        } catch (SchedulerException e){
+            logger.error("An exception occurred when collecting quartz metrics", e);
+        }
+    }
+
+    private void registerPendingJobs(JobExecutionContext context) throws SchedulerException {
+        groupJobsByClass(context.getScheduler().getJobKeys(GroupMatcher.anyJobGroup()))
```
[the remark this answers is not in the document]
Comment (comunacho) [reply webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzMetricsReportingJobListener.java:60]: This runs one query which targets indexed columns. For my setup, this is the prepared statement it runs for that:

`SELECT JOB_NAME, JOB_GROUP FROM QRTZ_JOB_DETAILS WHERE SCHED_NAME = 'TestScheduler' AND JOB_GROUP = ?`

So considering this runs on top of indexed columns, it shouldn't be really expensive.

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzMetricsReportingJobListener.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (aurambaj) [reply webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzMetricsReportingJobListener.java:53]: yes, let's do 1. for now since we don't have that many servers. If we're sending the size queue, we can just do an average.

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzMetricsReportingJobListener.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (aurambaj) [reply webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzMetricsReportingJobListener.java:60]: looks like it is fetching data so I'd say that's not ok. Potentially replacing with a similar `count` query and check the explain plan. But yeah as discussed maybe having a schedule task is better.

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    private void reportResults(PendingJob pendingJob){
+        meterRegistry.gauge("quartz.pending.jobs",
+                            Tags.of("jobClass", pendingJob.jobClass, "jobGroup", pendingJob.jobGroup),
+                            pendingJob.count);
+    }
+
+    List<PendingJob> fetchResults(){
+        return jdbcTemplate.query(
+            "SELECT job_class_name, job_group, COUNT(*) FROM QRTZ_JOB_DETAILS d GROUP BY d.job_class_name, d.job_group",
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java:39]: they don't have any abstraction on top of the quartz datasource to get the count? you can only list jobs?

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            (rs, num) -> new PendingJob(extractClassName(rs.getString(1)), rs.getString(2), rs.getLong(3))
+        );
+    }
+
+    private String extractClassName(String input) {
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java:44]: can you have a test for this?

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        String[] parts = input.split("\\.");
+        return parts[parts.length-1];
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java:46]: knit: indentation seems weird, do you apply intellij reformat?

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+@ConditionalOnExpression("'${spring.datasource.url:}'.toLowerCase().contains('mysql')")
+public class QuartzPendingJobsReportingTask {
+
+    @Autowired
+    JdbcTemplate jdbcTemplate;
+
+    @Autowired
+    MeterRegistry meterRegistry;
+
+    @Scheduled(fixedRateString = "${l10n.management.metrics.quartz-metrics.execution-rate}")
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java:26]: I suspect we can't use @ConfigurationProperties since it is an annotation?

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+import io.micrometer.core.instrument.search.MeterNotFoundException;
+import io.micrometer.core.instrument.search.RequiredSearch;
+
+import static com.box.l10n.mojito.quartz.QuartzConfig.DYNAMIC_GROUP_NAME;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+import static org.quartz.JobBuilder.newJob;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(classes = {Application.class},
```
Comment (aurambaj) [inline webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java:34]: can we either piggy back on ServiceTestBase or have a simple test that doesn't bootstrap the whole application? 

I don't know if they changed the implementation since then, but at the time we use to have slow builds because of different test contexts (and/or using dirties context). So unless having a test context that is really limited in scope and quick to start, I'd rather avoid re-introducing that.

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    @Autowired
+    Scheduler scheduler;
+
+    ConditionFactory await;
+    SecureRandom random = new SecureRandom();
+
+    @Before
+    public void setUp() throws Exception {
+        if (!scheduler.isStarted()){
+            scheduler.start();
```
Comment (aurambaj) [inline webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java:54]: in what case would this happen?

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        }
+        await = await().atMost(Duration.ofSeconds(5))
+                       .ignoreException(MeterNotFoundException.class);
```
Comment (aurambaj) [inline webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java:57]: what is this doing?

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    private void reportResults(PendingJob pendingJob){
+        meterRegistry.gauge("quartz.pending.jobs",
+                            Tags.of("jobClass", pendingJob.jobClass, "jobGroup", pendingJob.jobGroup),
+                            pendingJob.count);
+    }
+
+    List<PendingJob> fetchResults(){
+        return jdbcTemplate.query(
+            "SELECT job_class_name, job_group, COUNT(*) FROM QRTZ_JOB_DETAILS d GROUP BY d.job_class_name, d.job_group",
+            (rs, num) -> new PendingJob(extractClassName(rs.getString(1)), rs.getString(2), rs.getLong(3))
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java:40]: can you have a simple test just for this function?

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+                                .startAt(Date.from(ZonedDateTime.now().plusMinutes(5).toInstant()));
+        scheduler.scheduleJob(job, builder.build());
+
+        job = newJob(Test2Job.class).withIdentity("Test2Job", MY_GROUP).build();
+        builder = TriggerBuilder.newTrigger()
+                                .forJob(job)
+                                .startAt(Date.from(ZonedDateTime.now().plusMinutes(5).toInstant()));
+        scheduler.scheduleJob(job, builder.build());
+
+        await().until(() -> scheduler.isStarted());
```
Comment (aurambaj) [inline webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java:78]: related to @Before?

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java (hunk 6 above)
Comment (comunacho) [reply webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java:26]: Correct. The spring docs suggest using this format if a property is required.

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java (hunk 11 above)
Comment (comunacho) [reply webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java:78]: Correct

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java (hunk 8 above)
Comment (comunacho) [reply webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java:54]: I had a couple of failed situations when running my tests locally, so to avoid any failures it was better to have this as part of the setup.

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java (hunk 5 above)
Comment (comunacho) [reply webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java:46]: Yeah, its just intellij format

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java (hunk 3 above)
Comment (comunacho) [reply webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java:39]: I've spent some time on this yesterday and there's no abstraction for quartz that we could use. One option is to use the `JobStore` method `getJobKeys` along with `getJobGroupNames` but at the end of the day that'd be the same I had built initially.

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java (hunk 7 above)
Comment (comunacho) [reply webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java]: Updated

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java (hunk 10 above)
Comment (comunacho) [reply webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java:45]: Done

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java (hunk 9 above)
Comment (comunacho) [reply webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java:62]: This sets up the condition we'll use for awaitility later on. Given jobs can take some time to start and be processed, we wait at most 5 seconds and we ignore exceptions of Meters not found while we search for them (at t0 we might not have a specific meter given by the `RequiredSearch` conditions, but then at t1 the process that populates that Meter might have completed and populated that search)

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzMetricsReportingJobListener.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Component;
+
+import java.util.concurrent.TimeUnit;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tags;
+
+@ConditionalOnProperty(value = "l10n.management.metrics.quartz-metrics.enabled", havingValue = "true")
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzMetricsReportingJobListener.java:15]: knit: `l10n.management.metrics.quartz-metrics.enabled` --> `l10n.management.metrics.quartz.enabled` , no need to repeat metric, no?

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.util.List;
+
+import javax.sql.DataSource;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tags;
+
+@Component
+@ConditionalOnProperty(value = {"l10n.management.metrics.quartz-metrics.enabled",
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java:17]: aren't those two monitoring system independent? since it is opt-in, should it just check a single flag?

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java (hunk 5 above)
Comment (aurambaj) [reply webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java:51]: strange, mine add space around `-`

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java (hunk 9 above)
Comment (aurambaj) [reply webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java:62]: can you add this comment in the code?

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java (hunk 6 above)
Comment (aurambaj) [reply webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java:31]: sad, we don't get auto-complete on properties with this I beleive.

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+    @Test
+    public void testFetchResults() throws Exception {
+
+        JobDetail job;
+        TriggerBuilder<Trigger> builder;
+        int test1Count = random.nextInt(5) + 2;
+        int emptyCount = random.nextInt(5) + 2;
+
+        for (int i = 0; i < test1Count; i++) {
+            job = newJob(Test1Job.class).withIdentity("Test1Job_" + i, DYNAMIC_GROUP_NAME).build();
```
Comment (aurambaj) [inline webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java:150]: we've been using 
```
    @Rule
    public TestIdWatcher testIdWatcher = new TestIdWatcher();
```

to isolate test data between run and tests. Should probably used here too?

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+        await.untilAsserted(() -> assertThat(test1Gauges.gauge().value()).isEqualTo(test1Count));
+        await.untilAsserted(() -> assertThat(test2Gauges.gauge().value()).isEqualTo(emptyCount));
+    }
+
+    @Test
+    public void testFetchResults() throws Exception {
+
+        JobDetail job;
+        TriggerBuilder<Trigger> builder;
+        int test1Count = random.nextInt(5) + 2;
```
Comment (aurambaj) [inline webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java:146]: why random numbers here?

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+        assertThat(pendingJobs).extracting("jobClass")
+                               .containsExactlyInAnyOrder("QuartzPendingJobsReportingTaskTest$Test1Job", "EmptyTestJob");
+        assertThat(pendingJobs).extracting("jobGroup")
+                               .contains(DYNAMIC_GROUP_NAME);
+        assertThat(pendingJobs).extracting("count")
+                               .containsExactlyInAnyOrder((long) test1Count, (long) emptyCount);
+    }
+
+    @Test
+    public void testExtractClassName(){
```
Comment (aurambaj) [inline webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java:177]: indentation is weird again here.

`task.extractClassName("."))` may fail? i think that why I first asked for a test but then removed my comment because I supposed it is a edge case

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java (hunk 14 above)
Comment (comunacho) [reply webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java]: Done

File: webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java (hunk 13 above)
Comment (comunacho) [reply webapp/src/main/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTask.java]: Done

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java (hunk 15 above)
Comment (comunacho) [reply webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java]: Ahh, it was a leftover for when I wasn't acting upon the `QuartzPendingJobsReportingTask` directly. Updated

File: webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.search.MeterNotFoundException;
+import io.micrometer.core.instrument.search.RequiredSearch;
+
+import static com.box.l10n.mojito.quartz.QuartzConfig.DYNAMIC_GROUP_NAME;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+import static org.quartz.JobBuilder.newJob;
+
+@Profile("!disablescheduling")
```
Comment (aurambaj) [inline webapp/src/test/java/com/box/l10n/mojito/monitoring/QuartzPendingJobsReportingTaskTest.java:39]: oh, I wasn't clear, this is not meant to be on the test itself. But it is set on scheduled jobs. The profile is activated during testing so that those job don't run during testing.

**Comment by aurambaj** [comment]
Looks like the [issue with the checks ](https://github.com/box/mojito/issues/591)is happening again @jeeyi 

@comunacho the build seems to be failing: https://travis-ci.org/github/box/mojito/builds/701418962
